src/Learning/Parameters.jl

```diff
@@ -173,6 +173,51 @@ function backpropagate_tree!(Δ::AbstractMatrix{<:Real}, C::Vector{Node}, V::Abs
   return nothing
 end
 
+```
+  Marginalized version of backpropagate_tree.
+```
+function norm_backpropagate_tree!(marg_Δ::AbstractVector{<:Real}, C::Vector{Node}, marg_V::AbstractVector{<:Real})
+  marg_Δ .= 0.0
+  m = length(marg_Δ)
+  contains_indicators = false
+  # Columns: nodes
+  for i ∈ m:-1:1 # from root to leaves
+    N = C[i]
+    d = marg_Δ[i]
+    if isleaf(N) continue end
+    if issum(N)
+      for (j, c) ∈ enumerate(N.children)
+        if C[c] isa Indicator
+          # Indicator nodes are computed differently, since they may repeat (and so the structure is not
+          # tree-shaped). Compute in normal space so as to minimize numerical errors.
+          constains_indicators = true
+          marg_Δ[c] += N.weights[j] * exp(d)
+        else
+          marg_Δ[c] = log(N.weights[j]) + d
+        end
+      end
+    elseif isprod(N)
+      v = marg_V[i]
+      for c ∈ N.children
+        δ = d + v - marg_V[c]
+        isinvalid(δ)
```

Comment:
I think your code is based on an older version of `backpropagate`. Check the newer, improved and (hopefully) bug free current version here https://github.com/RenatoGeh/RPCircuits.jl/blob/a4cfef7c92c324aa4252f8c7ed834d1d6ac4e4af/src/Learning/Parameters.jl#L104.